### ISSUE TITLE
Not effective - CA2200: Rethrow to preserve stack details

### ISSUE THREAD

**Issue by LiangZugeng** [dotnet/roslyn-analyzers#6284]
### Analyzer

**Diagnostic ID**: [CA2200](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200): `Rethrow to preserve stack details `

**Version**: [SDK 7.0.100]

### Describe the bug

`dotnet build` or the build action in Visual Studio for Mac didn't show the warning of CA2200.

### Steps To Reproduce

1. With .NET 7 SDK installed, create a new console app using `dotnet new console -o TestConsoleApp`
2. Replace all content with the following code in Program.cs.

```
class TestsRethrow
{
    static void Main()
    {
        TestsRethrow testRethrow = new TestsRethrow();
        testRethrow.CatchException();
    }

    void CatchException()
    {
        try
        {
            CatchAndRethrowExplicitly();
        }
        catch (ArithmeticException e)
        {
            Console.WriteLine("Explicitly specified:{0}{1}",
               Environment.NewLine, e.StackTrace);
        }

        try
        {
            CatchAndRethrowImplicitly();
        }
        catch (ArithmeticException e)
        {
            Console.WriteLine("{0}Implicitly specified:{0}{1}",
               Environment.NewLine, e.StackTrace);
        }
    }

    void CatchAndRethrowExplicitly()
    {
        try
        {
            ThrowException();
        }
        catch (ArithmeticException e)
        {
            // Violates the rule.
            throw e;
        }
    }

    void CatchAndRethrowImplicitly()
    {
        try
        {
            ThrowException();
        }
        catch (ArithmeticException)
        {
            // Satisfies the rule.
            throw;
        }
    }

    void ThrowException()
    {
        throw new ArithmeticException("illegal expression");
    }
}
```

Content of the .csproj file
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>

```

### Expected behavior
Warning should be reported that CA2200 was found.

### Actual behavior
No warnings reported.

```
MSBuild version 17.4.0+18d5aef85 for .NET
  Determining projects to restore...
  Restored /Users/james/Projects/Temp/TestConsoleApp/TestConsoleApp.csproj (in 39 ms).
  TestConsoleApp -> /Users/james/Projects/Temp/TestConsoleApp/bin/Debug/net7.0/TestConsoleApp.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.64
```

## Additional context

.NET SDK: 7.0.100
OS: MacOS Ventura

**Comment by mavasani** [comment]
Interesting, we do have an existing test for this same scenario: https://github.com/dotnet/roslyn-analyzers/blob/2928b857198370bf518b8e5647188d2a055f5928/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/RethrowToPreserveStackDetailsTests.cs#L141-L166

But the analyzer is not firing on the same code snippet. I am wondering if this is due to a change in Operation tree shape that the analyzer expects. I'll investigate.

**Comment by mavasani** [comment]
The issue seems to be that starting C# 8, the operation tree shape for `throw ex` actually has an `IConversionOperation` wrapping the thrown exception operation if the type of `ex` is not `System.Exception`. This breaks few analyzers that do not handle conversions as operands of throw operation - I'll send a PR with the fix.

**Comment by mavasani** [comment]
No, this would go into .NET 8, especially given this is about false negatives and not false positives. You should be able to get the analyzer NuGet package with this fix from the dotnet public package feed.
